### DISCUSSION TITLE
Bitbucket access token seems to be available, but documentation and test not fixed

### DISCUSSION THREAD

**Discussion by Kuroki-g** [renovatebot/renovate#33008] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Bitbucket, official docker image (renovate/renovate)

### Please tell us more about your question or problem

I found a pull request merged that makes Bitbucket access tokens available (https://github.com/renovatebot/renovate/pull/16900), and issue (https://github.com/renovatebot/renovate/issues/14221)

And this is not reflected in the documentation.
https://docs.renovatebot.com/modules/platform/bitbucket/

On the other hand, the corresponding part of the test does not use access tokens (maybe still not be fixed).

https://github.com/renovatebot/renovate/blob/b05799d9c5079bcfc822c29e83ec0c461e771ef5/lib/workers/global/config/parse/env.spec.ts#L229-L235

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
If you test and find that it works as expected, please confirm to us and ideally raise a PR updating the docs to indicate that token can be supplied as an alternative to username/password. If you test and find that it doesn't work, please supply debug logs of the failure.

The test you highlighted looks like a copy/paste error and should be updated too.

**Reply by Kuroki-g** [reply]
There are some kinds of access token (workspace, project, repository) in bitbucket, and all token is able to use for bearer authentication according to the documentation.

- https://support.atlassian.com/bitbucket-cloud/docs/using-workspace-access-tokens/
- https://support.atlassian.com/bitbucket-cloud/docs/using-project-access-tokens/
- https://support.atlassian.com/bitbucket-cloud/docs/repository-access-tokens/

I tested using "Repository Access Token".
I followed bitbucket's procedure, and added following permission scopes.
- pullrequest
- pullrequest:write
- repository
- repository:write

And I added this token via environment variable, named "RENOVATE_TOKEN".

After execution of bitbucket pipelines Pull Request was raised:
![image](https://github.com/user-attachments/assets/f9db9ad3-b3ca-4820-a2a8-e27e31e4dcca)

The configuration that I used had been confirmed that this is valid using app password.

NOTE: The log is difficult to show because this is out private repository.

> ideally raise a PR updating the docs to indicate that token can be supplied as an alternative to username/password. If you test and find that it doesn't work, please supply debug logs of the failure.

I have created commit (https://github.com/renovatebot/renovate/commit/c57dcb3f921709f722fea9ef66a41a846178a2b1)
If my test is OK, I can raise PR.

> The test you highlighted looks like a copy/paste error and should be updated too.

I don't know much about snapshot-testing so this is difficult for me to do (only I found is following snapshots needs to be also fixed).

https://github.com/renovatebot/renovate/blob/0552d987b6d1abdabad36d2b36034ec6e06b1873/lib/workers/global/config/parse/__snapshots__/env.spec.ts.snap#L12

**Reply by viceice** [reply]
if the token can't access `/2.0/user` API, then you need to configure `username` otherwise renovate won't find it's issues / PRs correctly

**Reply by Kuroki-g** [reply]
@viceice Thanks you for the advice.

As you advices me, Repository Access Token has no support for `https://api.bitbucket.org/2.0/user`.
This is my curl result (https://developer.atlassian.com/cloud/bitbucket/rest/api-group-users/#api-group-users):
```
$ curl --request GET   --url 'https://api.bitbucket.org/2.0/user'   --header 'Authorization: Bearer {REPOSITORY_TOKEN}'   --header 'Accept: application/json'
{"type": "error", "error": {"message": "Token is invalid or not supported for this endpoint."}}
```

> then you need to configure username otherwise renovate won't find it's issues / PRs 

Excuse me, but could you please tell me why this is necessary?

About issues, Renovate not support bitbucket issues feature (https://docs.renovatebot.com/faq/#major-platform-features-not-supported-by-renovate, https://github.com/renovatebot/renovate/issues/20568), and Repository Access Token itself has no permission for issues (https://jira.atlassian.com/browse/BCLOUD-22603).

In case of PRs, it appears that the PR search was done correctly.
This is part of my pipelines log (deleted my repository information):

```
DEBUG: http cache: saving https://api.bitbucket.org/2.0/repositories/{ORGANIZATION}/{REPO_SLUG}/pullrequests?state=OPEN&state=MERGED&state=DECLINED&state=SUPERSEDED&fields=values.id%2Cvalues.title%2Cvalues.state%2Cvalues.links.commits.href%2Cvalues.summary.raw%2Cvalues.source.branch.name%2Cvalues.destination.branch.name%2Cvalues.reviewers.display_name%2Cvalues.reviewers.uuid%2Cvalues.reviewers.nickname%2Cvalues.reviewers.account_status%2Cvalues.created_on%2Cvalues.updated_on&q=&pagelen=50 (etag="23b75de350befe49788808b55d00eaed", lastModified=undefined) (repository={ORGANIZATION}/{REPO_SLUG})
DEBUG: Fetched 4 PRs to sync with cache (repository={ORGANIZATION}/{REPO_SLUG})
```

> need to configure username

[FYI] I tested username passing via `RENOVATE_USERNAME`, but the raised pull request was not used this field `username` in option.

**Reply by rarkins** [reply]
Why do you need to know why it's necessary?

**Reply by Kuroki-g** [reply]
Because I am not using Bitbucket issues. I am using Jira integration.

P.S. Actually I had forgotten about the Bitbucket issue tracker. My apologies.

**Reply by rarkins** [reply, marked as answer]
In initPlatform(), Renovate queries /user in order to learn its Bitbucket UUID: https://github.com/renovatebot/renovate/blob/3b9464c8fa4b74e32115a8e167b190bd846ad852/lib/modules/platform/bitbucket/index.ts#L95-L98

If an error is thrown, it only logs and does not abort: https://github.com/renovatebot/renovate/blob/3b9464c8fa4b74e32115a8e167b190bd846ad852/lib/modules/platform/bitbucket/index.ts#L99-L108

It seems that is _is_ ok to omit `username` from config. However if `/user` fails, Renovate doesn't know its UUID, and I'm not sure it will keep working well or not. It will mean that it no longer filters Issue and PRs using its own UUID, i.e. it will pull down the full list of each instead of just its own. My best guess is that it might still work though.

**Reply by Kuroki-g** [reply]
Thanks, I understand that UUID is used for filter of PR list.
It seems to be able to update dependencies without filters, so I will test it for a while.

/2.0/user API always fails, therefore I found not fully support Bitbucket Cloud Access Tokens.

Thank you so much for your assistance.
